src/main/src/main/java/org/geoserver/config/GeoServerDataDirectory.java

```diff
@@ -1211,14 +1211,27 @@ public URL locateResource(String uri) {
                         //GEOS-7025: Just get the path; don't try to create the file
                         file = Paths.toFile(root(), resource.path());
                     }
+                    
                     URL u = fileToUrlPreservingCqlTemplates(file);
+                    
                     if (url.getQuery() != null) {
                         try {
-                            return new URL(u.toString() + "?" + url.getQuery());
+                            u = new URL(u.toString() + "?" + url.getQuery());
                         } catch (MalformedURLException ex) {
+                            GeoServerPersister.LOGGER.log(Level.WARNING, "Error processing query string for resource with uri: " + uri, ex);
                             return null;
                         }
                     }
+                    
+                    if (url.getRef() != null) {
+                        try {
+                            u = new URL(u.toString() + "#" + url.getRef());
+                        } catch (MalformedURLException ex) {
+                            GeoServerPersister.LOGGER.log(Level.WARNING, "Error processing # fragment for resource with uri: " + uri, ex);
```

Comment:
I never quite know what level to log these warnings as, since this is a problem based on styling (under user control) I agree warning is appropriate.